transport/src/main/java/io/netty/channel/CoalescingBufferQueue.java

```diff
@@ -0,0 +1,181 @@
+ * and the consumer to dequeue differently sized buffers with the guarantee that when all the bytes in an added
+ * buffer are removed the promise associated with it will be completed when promise passed in during removal of the
+ * last byte completes.
+ *
+ * <p>This functionality is useful for aggregating or partitioning writes into fixed size buffers for framing protocols
+ * such as HTTP2.
+ */
+public final class CoalescingBufferQueue {
+
+    private static final InternalLogger logger = InternalLoggerFactory.getInstance(CoalescingBufferQueue.class);
+
+    private final ByteBufAllocator alloc;
+    private final ArrayDeque<Object> promisesAndBufs = new ArrayDeque<Object>();
+    private int readableBytes;
+
+    public CoalescingBufferQueue(ByteBufAllocator alloc) {
+        this.alloc = ObjectUtil.checkNotNull(alloc, "alloc");
+    }
+
+    /**
+     * Add a buffer to the end of the queue.
+     */
+    public void add(ByteBuf buf) {
+        promisesAndBufs.add(ObjectUtil.checkNotNull(buf, "buff"));
+        readableBytes += buf.readableBytes();
+    }
+
+    /**
+     * Add a buffer to the end of the queue and associate a promise with it that should be completed when
+     * all it's bytes have been consumed from the queue and written.
+     * @param buf to add to the tail of the queue
+     * @param promise to complete when all the bytes have been consumed and written, can be void.
+     */
+    public void add(ByteBuf buf, ChannelPromise promise) {
+        // buffers are added before promises so that we naturally 'consume' the entire buffer during removal
+        // before we complete it's promise.
+        ObjectUtil.checkNotNull(buf, "buf");
+        ObjectUtil.checkNotNull(promise, "promise");
+        promisesAndBufs.add(buf);
+        if (!promise.isVoid()) {
+            promisesAndBufs.add(promise);
+        }
+        readableBytes += buf.readableBytes();
+    }
+
+    /**
+     * Remove a {@link ByteBuf} from the queue with the specified number of bytes. Any added buffer who's bytes are
+     * fully consumed during removal will have it's promise completed when the passed aggregate {@link ChannelPromise}
+     * completes.
+     *
+     * @param bytes the maximum number of readable bytes in the returned {@link ByteBuf}.
+     * @param aggregatePromise used to aggregate the promises for the constituent buffers.
+     * @return a {@link ByteBuf} composed of the enqueued buffers.
+     */
+    public ByteBuf remove(int bytes, ChannelPromise aggregatePromise) {
+        if (bytes < 0) {
+            throw new IllegalArgumentException("bytes cannot be less than 0 : " + bytes);
+        }
+        ObjectUtil.checkNotNull(aggregatePromise, "aggregatePromise");
+
+        // Use isEmpty rather than readableBytes==0 as we may have a promise associated with an empty buffer.
+        if (promisesAndBufs.isEmpty()) {
+            return Unpooled.EMPTY_BUFFER;
+        }
+
+        ByteBuf toReturn = null;
```

Comment:
Having the utility I think will make the code a lot simpler.  Let me sketch out the idea a little more.  So if you had an inner class:
```java
private class BufComposer {
    private ByteBuf buf;
    private CompositeByteBuf composite;
    void add(ByteBuf newBuf) {
        if (buf == null) {
            buf = newBuf;
            return;
        }
        if (composite == null) {
            composite = alloc.compositeBuffer(promisesAndBufs.size() + 1);
            addToComposite(buf);
            buf = composite;
        }
        addToComposite(newBuf);
    }

    ByteBuf buf() {
        return buf;
    }

    private void addToComposite(ByteBuf newBuf) {
        composite.addComponent(newBuf);
        composite.writerIndex(composite.writerIndex() + newBuf.readableBytes);
    }
}
```

Then the work required in this loop gets a lot easier (and more understandable):

```java
BufComposer composer = new BufComposer();
    int originalBytes = bytes;
    for (;;) {
        Object entry = promisesAndBufs.poll();
        if (entry == null) {
            break;
        }
        if (entry instanceof ChannelPromise) {
            aggregatePromise.addListener(new ChannelPromiseNotifier((ChannelPromise) entry));
            continue;
        }
        ByteBuf entryBuffer = (ByteBuf) entry;
        if (entryBuffer.readableBytes() > bytes) {
            // Read a slice from the buffer and put the rest back in the queue.
            promisesAndBufs.addFirst(entryBuffer);
            entryBuffer = entryBuffer.readSlice(bytes).retain();
        }
        // Add the entry to the output.
        composer.add(entryBuffer);
        bytes -= entryBuffer.readableBytes();
    }
    readableBytes -= originalBytes - bytes;
    assert readableBytes >= 0;
    return composer.buf();
```

I may have missed some edge cases, but that's the general idea.